nullaway/src/test/java/com/uber/nullaway/VarargsTests.java

```diff
@@ -177,4 +207,211 @@ public void testSkipJetbrainsNotNullOnVarArgsFromThirdPartyJars() {
+            " }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "public class Test {",
+            "  public void testNullableVarargsArray(Object o1, Object o2, Object o3, @Nullable Object o4) {",
+            "    Utilities.takesNullableVarargsArray(o1, o2, o3, o4);",
+            "    Utilities.takesNullableVarargsArray(o1);", // Empty var args passed
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object) null);",
+            "    // BUG: Diagnostic contains: passing @Nullable parameter '(java.lang.Object[]) null' where @NonNull",
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object[]) null);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void typeUseAndDeclarationOnBoth() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Nullable.java",
+            "package com.uber;",
+            "import java.lang.annotation.ElementType;",
+            "import java.lang.annotation.Target;",
+            "@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE, ElementType.TYPE_USE})",
+            "public @interface Nullable {}")
+        .addSourceLines(
+            "Utilities.java",
+            "package com.uber;",
+            "public class Utilities {",
+            " public static String takesNullableVarargsArray(Object o, @Nullable Object @Nullable... others) {",
+            "  String s = o.toString() + \" \";",
+            "  // BUG: Diagnostic contains: enhanced-for expression others is @Nullable",
+            "  for (Object other : others) {",
+            "    s += (other == null) ? \"(null) \" : other.toString() + \" \";",
+            "  }",
+            "  return s;",
+            " }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "public class Test {",
+            "  public void testNullableVarargsArray(Object o1, Object o2, Object o3, @Nullable Object o4) {",
+            "    Utilities.takesNullableVarargsArray(o1, o2, o3, o4);",
+            "    Utilities.takesNullableVarargsArray(o1);", // Empty var args passed
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object) null);",
+            "    Utilities.takesNullableVarargsArray(o1, (java.lang.Object[]) null);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void varargsPassArrayAndOtherArgs() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.checkerframework.checker.nullness.qual.Nullable;",
+            "public class Test {",
+            "  static void takesVarargs(Object... args) {}",
+            "  static void test(Object o) {",
+            "    Object[] x = new Object[10];",
+            "    takesVarargs(x, o);",
```

Comment:
Shouldn't we complain here, since all the elements of `x` will be `null`? 

Guess not. It is this line:

| Annotation Type | Annotation Position | Mode | `@Nullable` varargs array | `@Nullable` array elements | `@Nullable` args at calls
------------------|---------------------|------|-----------------------|----------------------|-----------|
| declaration  |   -   |   standard | ❌ | ❌ | ✅

(I assume that is what applies with no annotations, even if the imported annotation is type use)

Do we just default to being permissive here to avoid new errors in existing codebases without having to go all the way to `legacy` mode?

Reply:
This is a separate issue, which is around checking of array initialization.  By default, thus far, we have been unsound and not checked for initialization of array elements at all.  Checking that array elements (for an array of `@NonNull` references) are initialized before they are used would be challenging and likely lead to many false positives.  The Checker Framework Nullness Checker also handles array initialization unsoundly by default; see [here](https://checkerframework.org/manual/#nullness-arrays).  We should probably document this in our JSpecify mode documentation.